frontend/src/metabase/querying/parameters/utils/query.ts

```diff
@@ -0,0 +1,223 @@
+}
+
+function applyFilterParameter(
+  query: Lib.Query,
+  stageIndex: number,
+  type: ParameterType,
+  target: StructuredParameterDimensionTarget,
+  value: ParameterValueOrArray,
+): Lib.Query {
+  const column = getParameterFilterColumn(query, stageIndex, target);
+  if (column == null) {
+    return query;
+  }
+
+  const filter = getParameterFilterClause(type, column, value);
+  if (filter == null) {
+    return query;
+  }
+
+  return Lib.filter(query, stageIndex, filter);
+}
+
+function getParameterFilterColumn(
+  query: Lib.Query,
+  stageIndex: number,
+  target: StructuredParameterDimensionTarget,
+) {
+  const columnRef = target[1];
+  const columns = Lib.filterableColumns(query, stageIndex);
+  const [columnIndex] = Lib.findColumnIndexesFromLegacyRefs(
+    query,
+    stageIndex,
+    columns,
+    [columnRef],
+  );
+  if (columnIndex < 0) {
+    return;
+  }
+
+  return columns[columnIndex];
+}
+
+function getParameterFilterClause(
+  type: ParameterType,
+  column: Lib.ColumnMetadata,
+  value: ParameterValueOrArray,
+) {
+  if (Lib.isDateOrDateTime(column)) {
+    return getDateParameterFilterClause(column, value);
+  }
+  if (Lib.isBoolean(column)) {
+    return getBooleanParameterFilterClause(type, column, value);
+  }
+  if (Lib.isNumeric(column)) {
+    return getNumberParameterFilterClause(type, column, value);
+  }
+  if (Lib.isString(column)) {
+    return getStringParameterFilterClause(type, column, value);
+  }
+}
+
+function getStringParameterFilterClause(
+  type: ParameterType,
+  column: Lib.ColumnMetadata,
+  value: ParameterValueOrArray,
+): Lib.ExpressionClause | undefined {
+  const values = normalizeStringParameterValue(value);
+  if (values.length === 0) {
+    return;
+  }
+
+  const operator = STRING_OPERATORS[type] ?? "=";
+  return Lib.stringFilterClause({
+    operator,
+    column,
+    values,
+    options: { caseSensitive: false },
+  });
+}
+
+function getNumberParameterFilterClause(
+  type: ParameterType,
+  column: Lib.ColumnMetadata,
+  value: ParameterValueOrArray,
+): Lib.ExpressionClause | undefined {
+  const values = normalizeNumberParameterValue(value);
+  if (values.length === 0) {
+    return;
+  }
+
+  const operator = NUMBER_OPERATORS[type] ?? "=";
+  return match({ operator, values })
+    .with(
+      { operator: P.union("=", "!=") },
+      { operator: P.union(">=", "<="), values: [P.number] },
+      { operator: "between", values: [P.number, P.number] },
+      () => Lib.numberFilterClause({ operator, column, values }),
+    )
+    .otherwise(() => undefined);
+}
+
+function getBooleanParameterFilterClause(
+  type: ParameterType,
+  column: Lib.ColumnMetadata,
+  value: ParameterValueOrArray,
+): Lib.ExpressionClause | undefined {
+  const values = normalizeBooleanParameterValue(value);
+  if (values.length === 0) {
+    return;
+  }
+
+  return Lib.booleanFilterClause({
+    operator: "=",
+    column,
+    values: values.map(item => (type === "string/!=" ? !item : item)),
```

Comment:
:x: Previously, if you had a dashboard filter set as "is not false", and then drilled down with it, "is not false" filter would be applied to a question. Now, "is not false" dashboard filter will be converted to "is true" during drill.

This gives different results if column has empty/null values.

Reply:
Fixed, please take a look

Reply:
There are no new unit tests for this - e.g. `Is trial is not true` and others. But no E2E tests, that's true.